### PR TITLE
Task #204616,#204887,Task #204683 camp registration api

### PR DESCRIPTION
**Pull request by GitVivekHub**
# I have ensured that following `Pull Request Checklist` is taken care of before sending this PR

links of task 

* [x] Code is formatted as per format decided
* [x] Updated acceptance criteria in tracker
* [x] Updated test cases in test-cases-tracker
* [x] Updated new API endpoint if any in common postman collection
* [x] Updated DB changes in db-tracker if any

### CODE REVIEW

File: src/src/camp/camp.service.ts [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+
+			// Check if learner_data is defined
+			if (
+				!newQdata ||
+				!Array.isArray(newQdata) ||
+				newQdata.length === 0
+			) {
+				return response.status(400).json({
+					success: false,
+					message: 'No learner data found or an error occurred.',
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:92]: Add language constant.

File: src/src/camp/camp.service.ts [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+			}
+
+			let learner_data = newQdata;
+
+			// Check if facilitator_id and learner_data have the same length
+			if (learner_id.length !== learner_data.length) {
+				return response.status(400).json({
+					success: false,
+					message:
+						'Learners do not belong to the corresponding facilitator.',
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:103]: Add language constant.

File: src/src/camp/camp.service.ts [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+				});
+			}
+
+			let create_group_object = {
+				name: "untitled",
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:108]: `name: "untitled"` change to
`name: "camp " + faciltator_camp_data?.data?.camps_aggregate?.aggregate?.count + 1`

File: src/src/camp/camp.service.ts [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+				type: body.type,
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:109]: `type: body.type,` change to `type: "camp",`

File: src/src/camp/camp.service.ts [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+				status: body.status,
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:110]: `status: body.status` change to `status: "not_registered`

File: src/src/camp/camp.service.ts [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+					[],
+					false,
+					[...this.returnFieldscamps, 'group_id', 'id'],
+				);
+			}
+     
+			let camp_id = createcampResponse?.camps?.id;
+
+			if (!createcampResponse?.camps?.group_id) {
+				await this.hasuraService.delete('groups', {
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:154]: add check `group_id` not undefined.

File: src/src/camp/camp.service.ts [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+				});
+
+				await this.hasuraService.delete('groups', {
+					id: group_id,
+				});
+
+				return response.status(500).json({
+					success: false,
+					data: {},
+					message: 'error occured during creating group user.',
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:197]: Add language constant.

File: src/src/camp/camp.service.ts [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+			let camp_id = createcampResponse?.camps?.id;
+
+			if (!createcampResponse?.camps?.group_id) {
+				await this.hasuraService.delete('groups', {
+					id: group_id,
+				});
+				return response.status(500).json({
+					success: false,
+					data: {},
+					message: 'Camp registration failed.',
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:160]: Add language constant.

File: src/src/camp/camp.service.ts [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+			});
+
+			const auditData = {
+				userId: facilitator_id,
+				mw_userid: facilitator_id,
+				context: 'camp.id',
+				context_id: camp_id,
+				oldData: {
+					group_id: group_id,
+					status: body?.status,
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:230]: `status: body.status` change to `status: "not_registered`

File: src/src/camp/camp.service.ts [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+					learner_id: [learner_id],
+				},
+				newData: {
+					group_id: group_id,
+					status: body?.status,
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:235]: `status: body.status` change to `status: "not_registered`

File: src/src/camp/camp.controller.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+	@Post('/camp-list')
+	@UseGuards(new AuthGuard())
+	campList(@Req() request: any,
+	@Body() body: any,
+	@Res() response: any,) {
+		return this.campService.campList(body, request,response);
+	}
+
+	@Post('/camp-details/:id')
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/camp/camp.controller.ts:33]: resolved

File: src/src/camp/camp.controller.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+import { AuthGuard } from 'src/modules/auth/auth.guard';
+
+@Controller('camp')
+export class CampController {
+    constructor(
+		private campService: CampService
+		
+	) {}
+ 
+  @Post('/register')
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/camp/camp.controller.ts:14]: resolved

File: src/src/camp/camp.service.ts [hunk 13]
```diff
@@ -0,0 +1,456 @@
...
+		'kit_received',
+		'kit_was_sufficient',
+		'kit_ratings',
+		'kit_feedback',
+		'group_id',
+	];
+
+	public returnFieldsgroupUsers = ['group_id', 'id'];
+
+	async registerCamp(body: any, request: any, response: any) {
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts:30]: resolved

File: src/src/camp/camp.service.ts (hunk 6 above)
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts:154]: resolved

File: src/src/camp/camp.service.ts (hunk 9 above)
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts:230]: resolved

File: src/src/camp/camp.service.ts (hunk 10 above)
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts:235]: resolved

File: src/src/camp/camp.service.ts [hunk 14]
```diff
@@ -0,0 +1,468 @@
...
+					[],
+					false,
+					[...this.returnFieldsgroupUsers, 'group_id', 'id'],
+				);
+			});
+
+			const auditData = {
+				userId: facilitator_id,
+				mw_userid: facilitator_id,
+				context: 'camp.id',
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:235]: `context: 'camp.id',` change to `context: 'camp',`

File: src/src/camp/camp.service.ts [hunk 15]
```diff
@@ -0,0 +1,468 @@
...
+		let member_type = "owner"
+		let status = "active"
+		
+
+		if (!facilitator_id) {
+			return resp.status(401).json({
+				success: false,
+				message: 'Unauthenticated User!',
+			});
+		}
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:327]: Remove. `Unauthenticated User`

File: src/src/camp/camp.service.ts [hunk 16]
```diff
@@ -0,0 +1,468 @@
...
+		const response = await this.hasuraServiceFromServices.getData(data);
+		const newQdata = response?.data?.camps;
+	
+		await Promise.all(newQdata?.map(async (item) => {
+			await Promise.all(item.group_users.map(async (userObj) => {
+			  let profilePhoto = userObj.user.profile_photo_1;
+			  if (profilePhoto?.[0]?.id !== undefined) {
+				const { success, data: fileData } = await this.uploadFileService.getDocumentById(userObj.user.profile_photo_1[0].id);
+		    	if (success && fileData?.fileUrl) {
+				  userObj.user.profile_photo_1[0].fileUrl = fileData.fileUrl;
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:374]: user need to json object `profile_photo_1`

File: src/src/camp/camp.service.ts [hunk 17]
```diff
@@ -0,0 +1,468 @@
...
+		let member_type = "owner"
+		let status = "active"
+		
+
+		if (!facilitator_id) {
+			return resp.status(401).json({
+				success: false,
+				message: 'Unauthenticated User!',
+			});
+		}
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:404]: remove

File: src/src/camp/camp.service.ts [hunk 18]
```diff
@@ -0,0 +1,468 @@
...
+
+		let qury = `query MyQuery {
+			camps(where: {id:{_eq:${camp_id}},group_users: {group: {academic_year_id: {_eq:${academic_year_id}}, program_id: {_eq:${program_id}}}, user: {}, member_type: {_eq:${member_type}}, status: {_eq:${status}}, user_id: {_eq:${facilitator_id}}}}) {
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:407]: remove `user: {}` from query

File: src/src/camp/camp.service.ts [hunk 19]
```diff
@@ -0,0 +1,468 @@
...
+			  }
+			}
+		  }
+		  
+		  
+		  `;
+		const data = { query: qury };
+		const response = await this.hasuraServiceFromServices.getData(data);
+		const newQdata = response?.data?.camps;
+
```
Comment (sagarkoshti1990) [inline src/src/camp/camp.service.ts:444]: check camp count and add validation camp not found if count is 0

File: src/src/user/user.service.ts [hunk 20]
```diff
@@ -1130,18 +1130,73 @@ export class UserService {
...
+		
 		for (let data of tempArray) {
 			if (oldData[data] !== newData[data]) {
 				storeOld[data] = oldData[data];
 				storeNew[data] = newData[data];
 			}
 		}
+	
+		
+
```
Comment (sagarkoshti1990) [inline src/src/user/user.service.ts:1150]: remove black rows

**Review by sagarkoshti1990 — CHANGES_REQUESTED** [review]

File: src/src/camp/camp.service.ts [hunk 21]
```diff
@@ -0,0 +1,456 @@
...
+				message: 'An error occurred during camp registration.',
+				error: error.message,
+			});
+		}
+	}
+
+	async checkCampInformation(id: any) {
+		let facilitator_id = id;
+		let query = `query MyQuery {
+            camps_aggregate(where: {group_users: {status: {_eq: "active"}, member_type: {_eq: "owner"}, user_id: {_eq:${facilitator_id}}}}){
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts]: resolved

File: src/src/camp/camp.service.ts [hunk 22]
```diff
@@ -0,0 +1,456 @@
...
+			) {
+				return response.status(401).json({
+					success: false,
+					data: {},
+					message: 'Camp registeration limit exceeded',
+				});
+			}
+
+			let query = `query MyQuery {
+            users(where: {_and: {program_beneficiaries: {_and: {user_id: {_in: [${learner_id}]}},status:{_eq:${beneficiary_status}}, facilitator_id: {_eq:${facilitator_id}}}}}){
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts]: resolved

File: src/src/camp/camp.service.ts (hunk 3 above)
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts:110]: resolved

File: src/src/camp/camp.service.ts (hunk 4 above)
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts]: resolved

File: src/src/camp/camp.service.ts (hunk 5 above)
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts]: resolved

File: src/src/camp/camp.service.ts (hunk 14 above)
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts:235]: resolved

File: src/src/camp/camp.service.ts (hunk 17 above)
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts:404]: resolved

File: src/src/camp/camp.service.ts (hunk 15 above)
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts:327]: resolved

File: src/src/camp/camp.service.ts (hunk 18 above)
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts:407]: resolved

File: src/src/camp/camp.service.ts (hunk 19 above)
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts:444]: resolved

File: src/src/user/user.service.ts (hunk 20 above)
Comment (GitVivekHub) [reply src/src/user/user.service.ts:1150]: resolved

File: src/src/camp/camp.service.ts (hunk 16 above)
Comment (GitVivekHub) [reply src/src/camp/camp.service.ts:374]: resolved

**Review by sagarkoshti1990 — APPROVED** [review]

**Pull request merged**